modules/util/src/main/scala/gpp/util/Zipper.scala

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gpp.util
+
+import cats._
+import cats.implicits._
+import cats.data.NonEmptyList
+import monocle.{ Prism, Traversal }
+import monocle.macros.Lenses
+
+/**
+  * Minimal zipper based on scalaz's implementation
+  * This is only meant for small collections. performance has not been optimized
+  */
+@Lenses
+final case class Zipper[A](lefts: List[A], focus: A, rights: List[A]) {
+
+  /**
+    * Modify the focus
+    */
+  def modify(f: A => A): Zipper[A] = copy(lefts, f(focus), rights)
+
+  /**
+    * Modify the focus
+    */
+  def modifyP(p: Prism[A, A]): Zipper[A] =
+    p.getOption(focus).map(f => copy(lefts, f, rights)).getOrElse(this)
+
+  /**
+    * Find and element and focus if successful
+    */
+  def findFocusP(p: PartialFunction[A, Boolean]): Option[Zipper[A]] =
+    findFocus(p.lift.andThen(_.getOrElse(false)))
+
+  /**
+    * How many items are in the zipper
+    */
+  def length: Int = lefts.length + 1 + rights.length
+
+  /**
+    * Find and element and focus if successful
+    */
+  def findFocus(p: A => Boolean): Option[Zipper[A]] =
+    if (p(focus)) this.some
+    else {
+      val indexLeft  = lefts.indexWhere(p)
+      val indexRight = rights.indexWhere(p)
+      if (indexLeft === -1 && indexRight === -1)
+        none
+      else if (indexLeft >= 0)
+        (lefts.splitAt(indexLeft) match {
+          case (Nil, h :: t)      =>
+            Zipper(Nil, h, t ::: focus :: rights)
+          case (x :: Nil, i :: l) =>
+            Zipper(List(x), i, l ::: focus :: rights)
+          case (x, i :: l)        =>
+            Zipper(x, i, (focus :: l.reverse).reverse ::: rights)
+          case _                  =>
+            this
+        }).some
+      else
+        (rights.splitAt(indexRight) match {
+          case (Nil, List(y)) =>
+            Zipper((focus :: lefts.reverse).reverse, y, Nil)
+          case (x, h :: t)    =>
+            Zipper((focus :: lefts.reverse).reverse ::: x, h, t)
+          case _              =>
+            this
+        }).some
+    }
+
+  def exists(p: A => Boolean): Boolean =
+    if (p(focus))
+      true
+    else
+      lefts.exists(p) || rights.exists(p)
+
+  def find(p: A => Boolean): Option[A] =
+    if (p(focus))
+      focus.some
+    else
+      lefts.find(p).orElse(rights.find(p))
+
+  def withFocus: Zipper[(A, Boolean)] =
+    Zipper(lefts.map((_, false)), (focus, true), rights.map((_, false)))
+
+  def toList: List[A] = (focus :: lefts.reverse).reverse ::: rights
```

Comment:
I think this might be wrong? `lefts` should be stored in reverse order to make it possible to move back quickly so it seems like it should be just

```scala
lefts.reverse ::: (focus :: rights)
```

Let's add a test case for `Zipper.fromNel(nel).toNel === nel` or something like that?

Reply:
This works but your solution should be faster as it does only one `reverse`. I'll add a test too